File: atlaskv-ts-sdk/src/api/WatchApi.ts
import { HttpClient } from "../utils/HttpClient.js";
import { WatchEvent } from "../models/WatchEvent.js";
import { Validation } from "../utils/Validation.js";

export interface WatchListener {
  /**
   * Invoked when a state mutation event (PUT, DELETE) is received.
   */
  onEvent(event: WatchEvent): void;
  /**
   * Invoked when an error is encountered in the watch stream.
   */
  onError?(error: Error): void;
  /**
   * Invoked when the connection to the watch SSE endpoint is successfully opened.
   */
  onConnected?(): void;
  /**
   * Invoked when the watch stream disconnects.
   */
  onDisconnected?(): void;
}

export class WatchSession implements AsyncIterable<WatchEvent> {
  private active = true;
  private readonly path: string;
  private readonly httpClient: HttpClient;
  private readonly listener?: WatchListener;

  private queue: WatchEvent[] = [];
  private pendingResolvers: {
    resolve: (value: IteratorResult<WatchEvent>) => void;
    reject: (err: unknown) => void;
  }[] = [];
  private abortController: AbortController | null = null;

  constructor(path: string, httpClient: HttpClient, listener?: WatchListener) {
    this.path = path;
    this.httpClient = httpClient;
    this.listener = listener;
  }

  public start(): void {
    this.runStreamLoop().catch((err) => {
      if (this.active) {
        this.listener?.onError?.(err);
        this.pushError(err);
      }
    });
  }

  public close(): void {
    if (this.active) {
      this.active = false;
      this.abortController?.abort();
      this.pushDone();
    }
  }

  public isActive(): boolean {
    return this.active;
  }

  private pushEvent(event: WatchEvent): void {
    if (this.pendingResolvers.length > 0) {
      const { resolve } = this.pendingResolvers.shift()!;
      resolve({ value: event, done: false });
    } else {
      this.queue.push(event);
    }
  }

  private pushError(err: Error): void {
    while (this.pendingResolvers.length > 0) {
      const { reject } = this.pendingResolvers.shift()!;
      reject(err);
    }
  }

  private pushDone(): void {
    while (this.pendingResolvers.length > 0) {
      const { resolve } = this.pendingResolvers.shift()!;
      resolve({ value: undefined as unknown as WatchEvent, done: true });
    }
  }

  public [Symbol.asyncIterator](): AsyncIterator<WatchEvent> {
    return {
      next: (): Promise<IteratorResult<WatchEvent>> => {
        if (this.queue.length > 0) {
          return Promise.resolve({ value: this.queue.shift()!, done: false });
        }
        if (!this.active) {
          return Promise.resolve({ value: undefined as unknown as WatchEvent, done: true });
        }
        return new Promise<IteratorResult<WatchEvent>>((resolve, reject) => {
          this.pendingResolvers.push({ resolve, reject });
        });
      },
      return: (): Promise<IteratorResult<WatchEvent>> => {
        this.close();
        return Promise.resolve({ value: undefined as unknown as WatchEvent, done: true });
      },
    };
  }

  private async runStreamLoop(): Promise<void> {
    let backoffMs = 500;

    while (this.active) {
      const url = `${this.httpClient.getActiveBaseUrl()}${this.path}`;
      const headers: Record<string, string> = {
        Accept: "text/event-stream",
      };

      this.abortController = new AbortController();

      try {
        const response = await fetch(url, {
          method: "GET",
          headers,
          signal: this.abortController.signal,
        });

        if (response.status === 200) {
          backoffMs = 500; // Reset backoff on success

          if (!response.body) {
            throw new Error("Response body is null");
          }

          // Fallback reading mechanism to support both getReader() and Symbol.asyncIterator
          let reader: { read(): Promise<{ value?: Uint8Array; done: boolean }> };
          if (typeof response.body.getReader === "function") {
            reader = response.body.getReader();
          } else if (Symbol.asyncIterator in response.body) {
            const iterator = (response.body as unknown as AsyncIterable<Uint8Array>)[Symbol.asyncIterator]();
            reader = {
              read: async () => {
                const next = await iterator.next();
                return { value: next.value, done: next.done ?? false };
              },
            };
          } else {
            throw new Error("Response body is not readable");
          }

          const decoder = new TextDecoder();
          let buffer = "";
          let currentEvent: string | null = null;

          while (this.active) {
            const { value, done } = await reader.read();
            if (done) break;

            buffer += decoder.decode(value, { stream: true });
            const lines = buffer.split("\n");
            buffer = lines.pop() ?? "";

            for (const line of lines) {
              if (!this.active) break;
              const trimmed = line.trim();
              if (trimmed.length === 0) {
                currentEvent = null; // Reset event boundary
                continue;
              }

              if (trimmed.startsWith("event:")) {
                currentEvent = trimmed.substring(6).trim();
              } else if (trimmed.startsWith("data:")) {
                const data = trimmed.substring(5).trim();

                if (currentEvent === "status" && data === "connected") {
                  this.listener?.onConnected?.();
                } else if (currentEvent === "message") {
                  try {
                    const event = JSON.parse(data) as WatchEvent;
                    this.listener?.onEvent(event);
                    this.pushEvent(event);
                  } catch (err) {
                    console.error("Failed to parse watch event JSON", data, err);
                  }
                } else if (currentEvent === "error") {
                  this.listener?.onError?.(new Error(`Server SSE error: ${data}`));
                }
              }
            }
          }
        } else if (response.status === 503) {
          console.warn(`Watch stream rejected with status 503 from ${url}`);
        } else {
          console.warn(`Watch stream received HTTP ${response.status}, retrying...`);
        }
      } catch (err) {
        if (!this.active) {
          break; // Normal close
        }
        const error = err as Error;
        console.debug(`Watch connection broken: ${error.message}, reconnecting...`);
      }

      if (this.active) {
        this.listener?.onDisconnected?.();
        await new Promise((resolve) => setTimeout(resolve, backoffMs));
        backoffMs = Math.min(backoffMs * 2, 10000);
      }
    }
  }
}

export class WatchApi {
  private readonly httpClient: HttpClient;

  constructor(httpClient: HttpClient) {
    this.httpClient = httpClient;
  }

  /**
   * Subscribes to real-time events for a single key.
   *
   * @param key the key to watch
   * @param listener callback for events and lifecycle changes
   * @returns a session object that can be closed to stop watching
   */
  public watch(key: string, listener?: WatchListener): WatchSession {
    Validation.validateKey(key);
    const session = new WatchSession(`/api/v1/watch/${key}`, this.httpClient, listener);
    session.start();
    return session;
  }

  /**
   * Subscribes to real-time events for all keys matching a prefix.
   *
   * @param prefix the prefix to watch
   * @param listener callback for events and lifecycle changes
   * @returns a session object that can be closed to stop watching
   */
  public watchPrefix(prefix: string, listener?: WatchListener): WatchSession {
    Validation.validatePrefix(prefix);
    const session = new WatchSession(`/api/v1/watch/prefix/${prefix}`, this.httpClient, listener);
    session.start();
    return session;
  }
}

File: atlaskv-java-sdk/src/main/java/com/atlaskv/sdk/api/WatchApi.java
package com.atlaskv.sdk.api;

import com.atlaskv.sdk.client.AtlasKVClient;
import com.atlaskv.sdk.models.WatchEvent;
import com.atlaskv.sdk.util.JsonUtil;
import com.atlaskv.sdk.util.ValidationUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.stream.Stream;

/**
 * Watch API client. Establishes real-time SSE streams to track mutations on key or prefixes.
 */
public final class WatchApi {

    private static final Logger LOG = LoggerFactory.getLogger(WatchApi.class);

    private final AtlasKVClient client;

    public WatchApi(AtlasKVClient client) {
        this.client = client;
    }

    /**
     * Subscribes to real-time events for a single key.
     *
     * @param key      the key to watch
     * @param listener callback for events and lifecycle changes
     * @return a session object that can be closed to stop watching
     */
    public WatchSession watch(String key, WatchListener listener) {
        ValidationUtil.validateKey(key);
        if (listener == null) {
            throw new IllegalArgumentException("WatchListener must not be null");
        }
        WatchSession session = new WatchSession("/api/v1/watch/" + key, listener);
        session.start();
        return session;
    }

    /**
     * Subscribes to real-time events for all keys matching a prefix.
     *
     * @param prefix   the prefix to watch
     * @param listener callback for events and lifecycle changes
     * @return a session object that can be closed to stop watching
     */
    public WatchSession watchPrefix(String prefix, WatchListener listener) {
        ValidationUtil.validatePrefix(prefix);
        if (listener == null) {
            throw new IllegalArgumentException("WatchListener must not be null");
        }
        WatchSession session = new WatchSession("/api/v1/watch/prefix/" + prefix, listener);
        session.start();
        return session;
    }

    /**
     * Represents an active watch subscription session.
     */
    public final class WatchSession implements AutoCloseable {

        private final String path;
        private final WatchListener listener;
        private final AtomicBoolean active = new AtomicBoolean(true);
        private volatile Thread streamThread;
        private volatile Stream<String> activeStream;

        private WatchSession(String path, WatchListener listener) {
            this.path = path;
            this.listener = listener;
        }

        private void start() {
            client.connectionPool().executorService().execute(this::runStreamLoop);
        }

        private void runStreamLoop() {
            this.streamThread = Thread.currentThread();
            int backoffMs = 500;

            while (active.get()) {
                URI activeUri = client.activeBaseUri();
                String rawPath = path;
                String fullPath = (activeUri.getPath() != null && !activeUri.getPath().isEmpty()
                        && !activeUri.getPath().equals("/")
                        ? activeUri.getPath().replaceAll("/+$", "") : "")
                        + (rawPath.startsWith("/") ? rawPath : "/" + rawPath);
                URI targetUri = URI.create(activeUri.getScheme() + "://" + activeUri.getAuthority() + fullPath);

                HttpRequest.Builder reqBuilder = HttpRequest.newBuilder()
                        .uri(targetUri)
                        .header("Accept", "text/event-stream")
                        .timeout(Duration.ofHours(24)) // Extremely long timeout for streaming
                        .GET();

                // Apply configured authentication (e.g. API key Bearer token)
                client.authentication().apply(reqBuilder);

                if (client.namespace() != null && !client.namespace().isBlank()) {
                    reqBuilder.header("X-Namespace", client.namespace());
                }

                HttpRequest request = reqBuilder.build();

                try {
                    LOG.debug("Establishing watch stream to {}", targetUri);
                    HttpResponse<Stream<String>> response = client.connectionPool().httpClient()
                            .send(request, HttpResponse.BodyHandlers.ofLines());

                    if (response.statusCode() == 200) {
                        backoffMs = 500; // Reset backoff on success
                        try (Stream<String> lines = response.body()) {
                            this.activeStream = lines;
                            if (active.get()) {
                                parseSseStream(lines);
                            }
                        } finally {
                            this.activeStream = null;
                        }
                    } else if (response.statusCode() == 503) {
                        // Leader changed or node unavailable, backoff and retry
                        LOG.warn("Watch stream rejected with status 503 from {}", targetUri);
                    } else if (response.statusCode() == 401 || response.statusCode() == 403) {
                        LOG.error("Watch stream rejected with auth error HTTP {}", response.statusCode());
                        listener.onError(new IOException("Watch authentication failed: HTTP " + response.statusCode()));
                        break;
                    } else {
                        LOG.warn("Watch stream received HTTP {}, retrying...", response.statusCode());
                    }
                } catch (IOException e) {
                    if (!active.get()) {
                        break; // Normal close
                    }
                    LOG.debug("Watch connection broken: {}, reconnecting...", e.getMessage());
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    break;
                } catch (Exception e) {
                    if (!active.get()) {
                        break;
                    }
                    LOG.error("Unexpected error in watch stream", e);
                    listener.onError(e);
                }

                // Disconnected callback prior to reconnection delay
                if (active.get()) {
                    listener.onDisconnected();
                    try {
                        Thread.sleep(backoffMs);
                        backoffMs = Math.min(backoffMs * 2, 10000); // Backoff up to 10s
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                        break;
                    }
                }
            }
        }

        private void parseSseStream(Stream<String> lines) {
            final String[] currentEvent = {null};
            lines.forEach(line -> {
                if (!active.get()) {
                    return;
                }
                String trimmed = line.trim();
                if (trimmed.isEmpty()) {
                    currentEvent[0] = null; // Reset event boundary
                    return;
                }

                if (trimmed.startsWith("event:")) {
                    currentEvent[0] = trimmed.substring(6).trim();
                } else if (trimmed.startsWith("data:")) {
                    String data = trimmed.substring(5).trim();
                    handleDataPayload(currentEvent[0], data);
                }
            });
        }

        private void handleDataPayload(String eventName, String data) {
            if ("status".equals(eventName) && "connected".equals(data)) {
                listener.onConnected();
            } else if ("message".equals(eventName)) {
                try {
                    WatchEvent event = JsonUtil.readValue(data, WatchEvent.class);
                    listener.onEvent(event);
                } catch (Exception e) {
                    LOG.error("Failed to parse watch event JSON: {}", data, e);
                }
            } else if ("error".equals(eventName)) {
                listener.onError(new IOException("Server SSE error: " + data));
            }
        }

        /**
         * Closes the watch stream and releases resources.
         */
        @Override
        public void close() {
            if (active.compareAndSet(true, false)) {
                LOG.debug("Closing watch session on path {}", path);
                Stream<String> s = activeStream;
                if (s != null) {
                    try {
                        s.close();
                    } catch (Exception e) {
                        LOG.debug("Error closing watch stream", e);
                    }
                }
                if (streamThread != null) {
                    streamThread.interrupt();
                }
            }
        }

        /**
         * Checks if the watch session is active.
         *
         * @return true if active
         */
        public boolean isActive() {
            return active.get();
        }
    }
}
